docs/src/apalache/fold.md

```diff
@@ -0,0 +1,188 @@
+
+  1. What are the semantics of fold operators?
+  1. How do I use these operators in Apalache?
+  1. Should I use folding or recursion?
+  1. Examples of common operators defined with folds
+
+### Syntax
+
+The syntax of the fold operators is as follows:
+
+```tla
+\* @type: ( (a, b) => a, a, Set(b) ) => a
+FoldSet( operator, base, set )
+
+\* @type: ( (a, b) => a, a, Seq(b) ) => a
+FoldSeq( operator, base, seq )
+```
+
+### Semantics of fold operators
+
+Folding refers to iterative application of a binary operator over a collection. Given an operator `Op`, a base value `b` and a collection of values `C`, the definition of folding `Op` over `C` starting with `b` depends on the type of the collection `C`.
+
+#### Semantics of `FoldSeq`
+
+In the case of folding over sequences, `C` is a sequence `<<a_1, ..., a_n>>`. Then, `FoldSeq( Op, b, C )` is defined as follows:
+
+  1. If `C` is empty, then `FoldSeq( Op, b, <<>> ) = b`, regardless of `Op`
+  1. If `C` is nonempty, we establish a recursive relation between folding over `C` and folding over `Tail(C)` in the following way: `FoldSeq( Op, b, C ) = FoldSeq( Op, Op(b, Head(C)), Tail(C) )`.
+
+#### Semantics of `FoldSet`
+
+In the case of folding over sequences, `C` is a set `{a_1, ..., a_n}`. Then, `FoldSet( Op, b, C )` is defined as follows:
+
+  1. If `C` is empty, then `FoldSeq( Op, b, {} ) = b`, regardless of `Op`
+  1. If `C` is nonempty, we establish a recursive relation between folding over `C` and folding over some subset of `C` in the following way: `FoldSeq( Op, b, C ) = FoldSeq( Op, Op(b, x), C \ {x} )`, where `x = CHOOSE y \in C: TRUE`. Note that Apalache selects `x` nondeterministically.
+
+Note that the above are definitions of a _left fold_ in the literature. Apalache does not implement a right fold.
+
+For example, if `C` is the sequence `<<x,y,z>>`, the result is equal to `Op( Op( Op(b, x), y), z)`. If `C = {x,y}`, the result is either `Op( Op(b, x), y)` or `Op( Op(b, y), x)`. Because the order of elements selected is nondeterministic in the case of sets, users should be careful, as the result is only uniquely defined in the case that the operator is both associative (`Op(Op(a,b),c) = Op(a,Op(b,c))`) and commutative (`Op(a,b) = Op(b,a)`). 
+
+For example, consider the operator `Op(p,q) == 2 * p + q`, which is noncommutative, and the set `S = {1,2,3}`. The value of `FoldSet(Op, 0, S)` depends on the order in which Apalache selects elements from S:
+
+
+| Order | FoldSet value |
+| --- | --- |
+| 1 -> 2 -> 3 | 11 | 
+| 1 -> 3 -> 2 | 12 | 
+| 2 -> 1 -> 3 | 13 | 
+| 2 -> 3 -> 1 | 15 | 
+| 3 -> 1 -> 2 | 16 | 
+| 3 -> 2 -> 1 | 17 | 
+
+Because Apalache chooses nondeterministically, all of the above results are possible outcomes, unlike in the case of the fixed-but-unknown order that would arise from a deterministic choice.
```

Comment:
Is this actually true? I thought that we implement `FoldSet` for some order that depends on the arena representation of a set. So we cannot guarantee that all possible orders will be explored, right?

Reply:
Only one order will be explored, but the point is that the user has no way of determining _which_ order, so they must consider all of them as possibilities.

Reply:
Right. The above explanation can be understood as there is a way to produce all possible iteration orders. This is the ambiguity of the word "non-determinism". So we should probably say "fixed but unknown order of iteration".

Reply:
"fixed but unknown order of iteration" is what happens with deterministic CHOOSE

Reply:
I've reworded the section to instead say something to the effect of "makes no guarantee of determinism". This way, we're not claiming nondeterminism, but are also in the clear if we choose to go with full determinism.